vertx-web/src/main/java/io/vertx/ext/web/handler/impl/FormLoginHandlerImpl.java

```diff
@@ -78,56 +82,51 @@ public FormLoginHandlerImpl(AuthenticationProvider authProvider, String username
   }
 
   @Override
-  public void handle(RoutingContext context) {
+  public void parseCredentials(RoutingContext context, Handler<AsyncResult<Credentials>> handler) {
     HttpServerRequest req = context.request();
     if (req.method() != HttpMethod.POST) {
-      context.fail(405); // Must be a POST
+      handler.handle(Future.failedFuture(BAD_METHOD)); // Must be a POST
     } else {
       if (!req.isExpectMultipart()) {
-        throw new IllegalStateException("HttpServerRequest should have setExpectMultipart set to true, but it is currently set to false.");
-      }
-      MultiMap params = req.formAttributes();
-      String username = params.get(usernameParam);
-      String password = params.get(passwordParam);
-      if (username == null || password == null) {
-        log.warn("No username or password provided in form - did you forget to include a BodyHandler?");
-        context.fail(400);
+        IllegalStateException ise = new IllegalStateException("HttpServerRequest should have setExpectMultipart set to true, but it is currently set to false.");
+        log.error("Failed to parse the credentials from request.", ise);
+        handler.handle(Future.failedFuture(ise));
       } else {
-        Session session = context.session();
-        UsernamePasswordCredentials authInfo = new UsernamePasswordCredentials(username, password);
-
-        authProvider.authenticate(authInfo, res -> {
-          if (res.succeeded()) {
-            User user = res.result();
-            context.setUser(user);
-            if (session != null) {
-              // the user has upgraded from unauthenticated to authenticated
-              // session should be upgraded as recommended by owasp
-              session.regenerateId();
+        MultiMap params = req.formAttributes();
+        String username = params.get(usernameParam);
+        String password = params.get(passwordParam);
+        if (username == null || password == null) {
+          log.warn("No username or password provided in form - did you forget to include a BodyHandler?");
```

Comment:
no more need for this log, as the `if` above can clearly identify if body handler was run or not

Reply:
Agree :)